notebooks/04_model_select_and_optimize/hypertune_spark_deep_dive.ipynb

```diff
@@ -966,42 +1505,42 @@
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td>spark</td>\n",
-       "      <td>0.920545</td>\n",
-       "      <td>0.729969</td>\n",
-       "      <td>0.341029</td>\n",
-       "      <td>0.321293</td>\n",
-       "      <td>56.728408</td>\n",
+       "      <td>0.916510</td>\n",
+       "      <td>0.728810</td>\n",
+       "      <td>0.350507</td>\n",
+       "      <td>0.327230</td>\n",
+       "      <td>41.124152</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td>hyperopt</td>\n",
-       "      <td>0.934265</td>\n",
-       "      <td>0.750182</td>\n",
-       "      <td>0.336724</td>\n",
-       "      <td>0.300553</td>\n",
-       "      <td>114.746458</td>\n",
+       "      <td>0.931035</td>\n",
+       "      <td>0.745581</td>\n",
+       "      <td>0.338503</td>\n",
+       "      <td>0.305381</td>\n",
+       "      <td>283.482648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td>sample</td>\n",
-       "      <td>0.965644</td>\n",
-       "      <td>0.761879</td>\n",
-       "      <td>0.269269</td>\n",
-       "      <td>0.252780</td>\n",
-       "      <td>221.232502</td>\n",
+       "      <td>0.971059</td>\n",
+       "      <td>0.767857</td>\n",
+       "      <td>0.265452</td>\n",
+       "      <td>0.244376</td>\n",
+       "      <td>580.297325</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "text/plain": [
        "   Approach      RMSE       MAE  Explained variance  R squared     Elapsed\n",
-       "0     spark  0.920545  0.729969            0.341029   0.321293   56.728408\n",
-       "0  hyperopt  0.934265  0.750182            0.336724   0.300553  114.746458\n",
-       "0    sample  0.965644  0.761879            0.269269   0.252780  221.232502"
+       "0     spark  0.916510  0.728810            0.350507   0.327230   41.124152\n",
+       "0  hyperopt  0.931035  0.745581            0.338503   0.305381  283.482648\n",
+       "0    sample  0.971059  0.767857            0.265452   0.244376  580.297325"
       ]
      },
-     "execution_count": 86,
+     "execution_count": 75,
      "metadata": {},
      "output_type": "execute_result"
     }
```

Comment:
So looking at the results, I get the explanation of the times. However, in terms of performance, basic grid search is better than TPE. I would love to understand this better.

In the paper, TPE outperforms manual and random hyperparameters, and even gaussian processes, however, in the notebook's results there is no improvement. 

We are saying: `This may be owing to the limited number of iterations used in the TPE searching algorithm. ` Can we test this?

Are there any examples with hyperopt where TPE outperforms manual grid search?

Adding @anargyri @eisber @gramhagen @loomlike for discussion.

Reply:
This was also what I have discussed with Andreas. I am doing the testing on Databricks where the size of datasets, # iterations, etc. are variables to test scalability and other things of our interest. Would love to have your thoughts on this.

Reply:
yeah, it's curious

Reply:
Agree. I can put those studies into the repo. Maybe after our setup of an Azure Databricks DevOps pipeline?

Reply:
The faster times for Spark may be due to the parallelism. It would be interesting to try the same tuning experiment with hyperopt on mongodb. 

However, even sequential hyperopt has an advantage because it is doing "intelligent" search of the space and not exhaustive grid search. So, there is a trade off between smart search and parallelism. For a small number of hyperparameters (2 in this notebook), parallelism will prevail. As the number of hyperparameters increases, I would expect hyperopt to catch up and at some point be faster than Spark grid search.

One way to show this would be to tune the `alpha` and `nonnegative` hyperparameters of ALS, so that you search a 4d space of hyperparameters. We would probably see better times for hyperopt relative to Spark.

Reply:
what puzzles me is that even though

> it is doing "intelligent" search of the space and not exhaustive grid search.

The metrics are not better in the "intelligent search"

Reply:
They should not be better if exhaustive search was done completely. It is better in cases when it is not possible to search the hyperparameter space in an exhaustive way.

Reply:
in this case we are not doing a complete search

Reply:
Actually it is not clear if ALS finds a good set of hyperparameters, because we don't see the precision in the output. 
Also note that we optimize precision but we report regression metrics. @yueguoguo shouldn't we report precision too?

Reply:
It is hard if not impossible to find a "complete search" without a prior knowledge of the parameter range. Regardless random search or grid search, the parameter space is predefined by human and the two approaches just find the optimal ones in different ways - the random search can benefit from some "intelligent" method (e.g.., Bayesian method, early-stop policy, etc.) to reduce the # of iterations, while the second does nothing "intelligent" but to finish the whole iterations and pick the optimal from all.

I think I need to rephrase a little bit in the notebook that this is NOT for comparison. Instead, it is just for providing users with information about how different approaches work and how to choose by a given criterion. A scalable version of the notebook (on Databricks) makes more sense for comparison. What do you think?

Reply:
> Also note that we optimize precision but we report regression metrics. @yueguoguo shouldn't we report precision too?

Precision is meant to demonstrate that we can customize Spark constructs to allow any metrics of interest. It is not for comparison. I should have made it more clear in the notebook text.

Reply:
I think it becomes a bit confusing because you call the dataframe "comparison" and many people will focus on the table and view it as some type of comparison. 
IMO I would like to avoid giving the impression that if you tune by optimizing precision you also optimize RMSE, MAE etc. as a result. We know that this is not correct, but some people may not be aware.

Reply:
Good point. I will rephrase this part to make it sound non-opinionated.